Chapter: Background / REVIEW (#27)
ysndr/MSc-Thesis: chapter/background.md

# Background

This thesis illustrates an approach of implementing a language server for the Nickel language which communicates with its clients, i.e. editors, over the open Language Server Protocol (in the following abbreviated as *LSP*).
The current chapter provides the background on the technological details of the project.
As the work presented aims to be transferable to other languages using the same methods, this chapter will provide the means to distinguish the nickel specific implementation details.

The primary technology built upon in this thesis is the language server protocol.
The first part of this chapter introduces the LSP, its rationale and improvements over classical approaches, technical capabilities and protocol details. 
The second part is dedicated to Nickel, elaborating on the context and use-cases of the language followed by an inspection of the technical features of Nickel.

## Language Server Protocol

> 

Author: ```suggestion

The Language Server Protocol is a JSON-RPC based communication specification comprising an LSP client (i.e. editors) and server (also called language server for simplicity).
The LSP decouples the development of clients and servers, allowing developers to focus on either side.
The LSP defines several capabilities -- standardised functions remotely executed by the language server.
LSP Clients are often implemented as editor extensions facilitating abstraction libraries helping with the interaction with the protocol and editor interface.
Language Servers analyse source code sent by the client and may implement any number of 
capabilities relevant to the language.
Since the LSP is both language and editor independent, the same server implementation can serve all LSP compliant clients eliminating the need to redundantly recreate the same code intelligence for every editor.
  
```

> Language servers are today's standard of integrating support for programming languages into code editors.

Reviewer: First define, before telling that they are standard or done by X. 

Language Servers are XXXXXX

Do you need to define / make a difference between "language server" and "language server protocol"? clarify.

Author: https://github.com/ysndr/MSc-Thesis/pull/27/files#r855145634

Initially developed by Microsoft for the use with their polyglot editor Visual Studio Code^[https://code.visualstudio.com/] before being released to the public in 2016 by Microsoft, RedHat and Codeenvy, the LSP decouples language analysis and provision of IDE-like features from the editor.
Developed under open source license on GitHub^[https://github.com/microsoft/language-server-protocol/], the protocol allows developers of editors and languages to work independently on the support for new languages.

> If supported by both server and client, the LSP now supports more than 24 language features^[https://microsoft.github.io/language-server-protocol/specifications/specification-current/] including code completion, code navigation facilities, contextual information such as types or documentation, formatting, and more.

Reviewer: interesting. add a recap table of those 24 features? then you can later explain on which ones you focus and why.

### JSON-RPC

> JSON-RPC (v2) [@json-rpc] is a JSON based lightweight transport independent remote procedure call [@rpc] protocol used by the LSP to communicate between a language server and a client.

Reviewer: First say "LSP uses JSON-RPC" (the transition with before). Then you proceed by explaining what JSON-RPC is

> JSON-RPC (v2) [@json-rpc] is a JSON based lightweight transport independent remote procedure call [@rpc] protocol used by the LSP to communicate between a language server and a client.

Author: ```suggestion
the LSP uses JSON-RPC to communicate between a language server and a client.
JSON-RPC (v2) [@json-rpc] is a JSON based lightweight transport independent remote procedure call [@rpc] protocol. 
```

> RPC is a well known paradigm that allows clients to virtually invoke a method at a connected process.

Reviewer: "well known " subjective, we don't care. Say what it is, as a fact.

Author: ```suggestion
RPC is a paradigm that allows clients to virtually invoke a method at a connected process.
```

In fact, JSON-RPC only specifies a message protocol, leaving the choice of the transport method to the application. 

All messages refer to a `method` and a set of `parameters`.
Servers are expected to perform the same procedure associated with the requested `method` at any time.
In return, clients have to follow the calling conventions for the requested method.
Typically, messages are synchronous, i.e., the client awaits a result associated to the method and its parameters before it continues its execution of the calling environment.
Hence, requests are marked with an `id` which is included in the result/error message necessarily sent by the server.
If the client does not require a response, it can omit the `id` field.
The message is then interpreted as a *notification*, which servers cannot respond to.

Part of the JSON-RPC specification is the ability for clients to batch requests and send a list of request or notification objects.
In this case, the server should respond with a list of results matching each request, yet is free to process requests concurrently.

### Commands and Notifications

The LSP builds on top of the JSON-RPC protocol described in the previous subsection.

> In total the LSP defines 33 [@lsp] "language features", i.e., source code related capabilities.

Reviewer: idem, add reference list here or in appendix

> The LSP builds on top of the JSON-RPC protocol described in the previous subsection.
> In total the LSP defines 33 [@lsp] "language features", i.e., source code related capabilities.
> In addition, the LSP specifies different capabilities to the server to control the editor.
> For instance, servers may instruct clients to show notifications or progress bars or open documents.
> Similarly, the client has multiple ways of notifying the server of file changes, including renaming r deletion of files.

Author: ```suggestion
The LSP builds on top of the JSON-RPC protocol described in the previous subsection.
It defines four sets of commands: 

The largest group are commands that are relevant in the scope of the currently opened document, e.g. autocompletion, refactoring, inline values and more.
In total the LSP defines 33 [@lsp] of such "Language Features".
Editors will notify the server about file changes and client side interaction, i.e., opening, closing and renaming files using "Document Synchronization" methods.
While most commands are defined in the document scope, i.e., a single actively edited file, the LSP allows clients to communicate changes to files in the opened project.
This so called workspace comprises on or more root folders managed by the editor and all files contained in them.
"Workspace Features" allow the server to intercept file creation, renaming or deletion to make changes to existing sources in other files.
Use cases of these features include updating import paths, changing class names and other automations that are not necessary local to a single file.
In addition, the LSP specifies so called "Window Features" which allow the server to control parts of the user interface of the connected editor.
For instance, servers may instruct clients to show notifications and progress bars or open files.
```

> This thesis aims to implement a fundamental set of capabilities.

Reviewer: not a sentence for here. Consider that this section is self-contained, it should be readable independently from the rest of the thesis.

> This thesis aims to implement a fundamental set of capabilities.
> The chosen capabilities are based on those identified as "key methods" by the authors of langserver [@langserver], specifically:

Author: ```suggestion

### Description of Key Commands

the authors of langserver.org [@langserver] identified six "key methods" of the LSP.
The methods represent a fundamental set of capabilities:
```

> 1. Code completion

Reviewer: very good. move to "## Language Server Protocol"
here we have the ####

Author: I don't understand,
I moved it into a new dedicated subsection but having all methods as direct subsections of `## LSP` implies to me that these are the only methods

   Suggest identifiers, methods or values at the cursor position.
2. Hover information
   Present additional information about an item under the cursor, i.e., types, contracts and documentation.
3. Jump to definition
   Find and jump to the definition of a local variable or identifier.
4. Find references
   List all usages of a defined variable.
5. Workspace/Document symbols
   List all variables in a workspace or document.
6. Diagnostics
   Analyze source code, i.e., parse and type check and notify the LSP Client if errors arise.

#### Code Completion

> 

Author: ```suggestion

*RPC Method: [`textDocument/Completion`](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_completion)*

```

This feature allows users to request a suggested identifier of variables or methods, concrete values or larger templates of generated code to be inserted at the position of the cursor.
The completion can be invoked manually or upon entering language defined trigger characters, such as `.`, `::` or `->`.
The completion request contains the current cursor position, allowing the language server to resolve contextual information based on an internal representation of the document.

> In the example ([@fig:lsp-capability-complete]) the completion feature suggests related identifiers for the incomplete function call "`pr`*`int`*".

Reviewer: add the JSON-RPC method names for all #### levels.

> ![](examples/complete.png){#fig:lsp-capability-hover caption="Completion options resolved by the Python language server in Visual Studio Code"}

Reviewer: fundamental limitation of reviewing a PR instead of reviewing a PDF, I don't see the figure inline.

#### Hover

> 

Author: ```suggestion

*RPC Method: [`textDocument/hover`](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_hover)*

```

Hover requests are issued by editors when the user rests their mouse cursor on text in an opened document or issues a designated command in editors without mouse support.
If the language server has indexed any information corresponding to the position, it can generate a report using plain text and code elements, which are then rendered by the editor.
Language servers typically use this to communicate type-signatures or documentation.
An example can be seen in [@fig:lsp-capability-hover].

![](examples/hover.png){#fig:lsp-capability-hover caption="Hover information displayed by the Python language server in Visual Studio Code"}

#### Jump to Definition

> 

Author: ```suggestion

*RPC Method: [`textDocument/definition`](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_definition)*

```

The LSP allows users to navigate their code by the means of symbols by finding the definition of a requested symbol.
Symbols can be for instance variable names or function calls.
As seen in [@fig:lsp-capability-definition], editors can use the available information to enrich hover overlays with the hovered element's definition.

![](examples/definition.png){#fig:lsp-capability-hover caption="Hover information enriched with the definition. Python language server in Visual Studio Code"}

#### Find References

> 

Author: ```suggestion

*RPC Method: [`textDocument/references`](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_references)*

```

Finding references is the inverse operation to the previously discussed *Jump to Definition* ([cf. @sec:jump-to-definition]).
For a given symbol definition, for example variable, function, function argument or record field the LSP provides all usages of the symbol allowing users to inspect or jump to the referencing code position.

![](examples/references.png){#fig:lsp-capability-hover caption="Listing of all references to the method "hello". Python language server in Visual Studio Code"}

#### Workspace/Document symbols

> 

Author: ```suggestion

*RPC Method: [`workspace_symbol`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_symbol) or [`textDocument_documentSymbol`](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_documentSymbol)*

```

> The symbols capability allows language servers to expose a list if symbols declared in the open document or workspace.

Reviewer: clearly define "document" and "workspace" in this context first.

Author: ```suggestion
The symbols capability is defined as both a "Language Feature" and "Workspace Feature" which mainly differ in the scope they represent.
The `textDocument/documentSymbol` command lists symbols solely found in the currently opened file, while the `workspace/symbol` takes into account all files in the opened set of folders.
```

> The granularity of the listed items is determined by the server.

Author: ```suggestion
The granularity of the listed items is determined by the server and possibly different for either scope.
For instance, document symbols could be as detailed as listing any kind of method or property found in the document, while workspace symbols take visibility rules into account which might expose public entities only.
```

Yet, it is far less performant.
Unlike incremental processing (which updates only the affected portion of its internal structure), the smallest changes, including adding or removing lines effect the _reprocessing of the entire file_.
While sufficient for small languages and codebases, non-incremental processing quickly becomes a performance bottleneck.

For code analysis LSP implementers have to decide between *lazy* or *greedy* approaches for processing files and answering requests.
Dominantly greedy implementations resolve most available information during the indexing of the file.
The server can then utilize this model to answer requests using mere lookups.
This stands in contrast to lazy approaches where only minimal local information is resolved during the indexing.
Requests invoke an ad-hoc resolution the results of which may be memoized for future requests.
Lazy resolution is more prevalent in conjunction with incremental indexing, since it further reduces the work associated with file changes.
This is essential in complex languages that would otherwise perform a great amount of redundant work.

> ## Configuration programming languages

Author: ```suggestion
## Text based Configuration
```

> Nickel [@nickel], the language targeted by the language server detailed in this thesis, defines itself as "configuration language" used to automize the generation of static configuration files.

Reviewer: remove. First define "configuration language".

Add a citation to a reference publication / book on the topic. Ina background / RW section all sections start with a citation.

Author: ```suggestion
```

> Static configuration languages such as XML[@xml], JSON[@json], or YAML[@yaml] are language specifications defining how to textually represent structural data used to configure parameters of a system^[some of the named languages may have been designed as a data interchange format which is absolutely compatible with also acting as a configuration language].

Reviewer: Is "dynamic configuration language" a thing? if not remove "static"

> Static configuration languages such as XML[@xml], JSON[@json], or YAML[@yaml] are language specifications defining how to textually represent structural data used to configure parameters of a system^[some of the named languages may have been designed as a data interchange format which is absolutely compatible with also acting as a configuration language].

Reviewer: You want to have a dedicated ### for the most notable config languages before Nickel (at the very least YAML, but possibly XML and others)

> Static configuration languages such as XML[@xml], JSON[@json], or YAML[@yaml] are language specifications defining how to textually represent structural data used to configure parameters of a system^[some of the named languages may have been designed as a data interchange format which is absolutely compatible with also acting as a configuration language].

Author: ```suggestion 
Configuration languages such as XML[@xml], JSON[@json], or YAML[@yaml] are textual representations of structural data used to configure parameters of a system^[some of the named languages may have been designed as a data interchange format which is absolutely compatible with also acting as a configuration language].
```

> Applications of configuration languages are ubiquitous especially in the vicinity of software development. While XML and JSON are often used by package managers [@npm, @maven, @composer], YAML is a popular choice for complex configurations such as CI/CD pipelines [@travis, @ghaction, @gitlab-runner] or machine configurations in software defined networks such as Kubernetes and docker compose.
> 
> Such static formats are used due to some significant advantages compared to other formats.
> Most strikingly, the textual representation allows inspection of a configuration without the need of a separate tool but a text editor and be version controlled using VCS software like Git.
> For software configuration this is well understood as being preferable over databases or other binary formats. Linux service configurations (files in `/etc`) and MacOS `*.plist` files which can be serialized as XML or a JSON-like format, especially exemplify that claim.
> 
> Yet, despite these formats being simple to parse and widely supported [@json], their static nature rules out any dynamic content such as generated fields, functions and the possibility to factorize and reuse.
> Moreover, content validation has to be developed separately, which led to the design of complementary schema specification languages like json-schema [@json-schema] or XSD [@xsd].
> 
> These qualities require an evaluated language.
> In fact, some applications make heavy use of config files written in the native programming language which gives them access to language features and existing analysis tools.
> Examples include JavaScript frameworks such as webpack [@webpack] or Vue [@vue] and python package management using `setuptools`[@setuptools].
> 
> Despite this, not all languages serve as a configuration language, e.g. compiled languages and some domains require language agnostic formats.
> For particularly complex products, both language independence and advanced features are desirable.
> Alternatively to generating configurations using high level languages, this demand is addressed by more domain specific languages.
> Dhall [@dhall], Cue [@cue] or jsonnet [@jsonnet] are such domain specific languages (DSL), that offer varying support for string interpolation, (strict) typing, functions and validation.

Author: ```suggestion
### Applications of Configuration Languages 

Applications of configuration languages are ubiquitous especially in the vicinity of software development.
While XML and JSON are often used by package managers [@npm, @maven, @composer], YAML is a popular choice for complex configurations such as CI/CD pipelines [@travis, @ghaction, @gitlab-runner] or machine configurations in software defined networks such as Kubernetes[@kubernetes] and docker compose [@docker-compose].

Such formats are used due to some significant advantages compared to other formats.
Most strikingly, the textual representation allows inspection of a configuration without the need of a separate tool but a text editor.
Moreover textual configuration can be version controlled using VCS software like Git.
For software configuration this preferred over databases or other binary formats.
Linux service configurations (files in `/etc`) and MacOS `*.plist` files which can be serialized as XML or a JSON-like format, especially exemplify that claim.

### Configuration *Programming* Languages

Despite the above-mentioned formats being simple to parse and widely supported [@json], their static nature rules out any dynamic content such as generated fields, functions and the possibility to factorize and reuse.
Moreover, content validation has to be developed separately, which led to the design of complementary schema specification languages like json-schema [@json-schema] or XSD [@xsd].

These qualities require an evaluated language.
In fact, some applications make heavy use of config files written in the native programming language which gives them access to language features and existing analysis tools.
Examples include JavaScript frameworks such as webpack [@webpack] or Vue [@vue] and python package management using `setuptools`[@setuptools].

Despite this, not all languages serve as a configuration language, e.g. compiled languages. 
For particularly complex applications, both language independence and advanced features are desirable.
Alternatively to generating configurations using high level languages, this demand is addressed by domain specific languages.
Dhall [@dhall], Cue [@cue] or jsonnet [@jsonnet] are such domain specific languages (DSL), that offer varying support for string interpolation, (strict) typing, functions and validation.
Most of these languages are used as a templating system which means a configuration file in a more portable format is generated using an evaluation of the more expressive configuration source.
The added expressiveness manifests in the ability to evaluate expression and the availability of imports, variables and functions.
These features allow to refactor and simplify repetitive configuration files.
```

### Infrastructure as Code

> A prime example for the application of configuration languages are IaaS^[Infrastructure as a Service] products.

Reviewer: Do all configuration languages feat under IaC? Is IaC a superset or a subset of configuration languages?

Make a transition for the reader.

> A prime example for the application of configuration languages are IaaS^[Infrastructure as a Service] products.
> These solutions offer great flexibility with regard to resource provision (computing, storage, load balancing, etc.), network setup and scaling of (virtual) servers.
> Although the primary interaction with those systems is imperative, maintaining entire applications' or company's environments manually comes with obvious drawbacks.

Author: ```suggestion
The shift to an increasing application of IaaS^[Infrastructure as a Service] products started the desire for declarative machine configuration in a bid to simplify the setup and reproducibility of such systems.
This configuration based setup of infrastructure is commonly summarized as infrastructure as code or IaC.
As the name suggests, IaC puts cloud configuration closer to the domain of software development [@IaC-book].

In general IaaS solutions offer great flexibility with regard to resource provision (computing, storage, load balancing, etc.), network setup and scaling of (virtual) servers.
However, since the primary interaction with those systems is imperative and based on command line or web interfaces, maintaining entire applications' or company's environments manually comes with obvious drawbacks.
```

Changing and undoing changes to existing networks requires intricate knowledge about its topology which in turn has to be meticulously documented.
Undocumented modification pose a significant risk for *config drift* which is particularly difficult to undo imperatively.
Beyond that, interacting with a system through its imperative interfaces demands qualified skills of specialized engineers.

> The concept of "Infrastructure as Code" (*IaC*) serves the DevOps principles.
> IaC tools help to overcome the need for dedicated teams for *Dev*elopment and *Op*erations by allowing to declaratively specify the dependencies, topology and virtual resources.
> Optimally, different environments for testing, staging and production can be derived from a common base and changes to configurations are atomic.
> As an additional benefit, configuration code is subject to common software engineering tooling;
> It can be statically analyzed, refactored and version controlled to ensure reproducibility.

Author: ```suggestion
The concept of "Infrastructure as Code" (*IaC*) serves the DevOps principles.
IaC tools help to overcome the need for dedicated teams for *Dev*elopment and *Op*erations by allowing to declaratively specify the dependencies, topology and virtual resources.
Optimally, different environments for testing, staging and production can be derived from a common base and changes to configurations are atomic.
As an additional benefit, configuration code is subject to common software engineering tooling;
It can be statically analyzed, refactored and version controlled to ensure reproducibility.
A subset of IaC is focused on largely declarative configuration based on configuration files that are interpreted and "converted" into imperative platform dependent instructions. 
```

        };
      };
      networking.firewall.allowedTCPPorts = [ 80 ];
      
      # Instance can override default deployment options 
      deployment.ec2.instanceType = "t1.medium";
    };
}
```

Similarly, tools like Terraform[@terraform], or Chef[@chef] use their own DSLs and integrate with most major cloud providers.
The popularity of these products^[https://trends.google.com/trends/explore?date=2012-01-01%202022-01-01&q=%2Fg%2F11g6bg27fp,CloudFormation], beyond all, highlights the importance of expressive configuration formats and their industry value.

Finally, descriptive data formats for cloud configurations allow mitigating security risks through static analysis.
Yet, as recently as spring 2020 and still more than a year later dossiers of Palo Alto Networks' security department Unit 42 [@pa2020H1, ps2021H2] show that a majority of public projects uses insecure configurations.
This suggests that techniques[@aws-cloud-formation-security-tests] to automatically check templates are not actively employed, and points out the importance of evaluated configuration languages which can implement passive approaches to security analysis.

### Nickel

> The Nickel[@nickel] language is a configuration programming language (cf. [@sec:configuration-programming-languages]) with the aims of providing composable, verifiable and validatable configuration files.

Reviewer: "configuration programming language" do you want to make a difference between "config language" and "configuration programming language"? if yes, first define both terms, and then say "Nickel is a configuration programming language because XXXX"

Author: ```suggestion
The Nickel[@nickel] language is a configuration programming language as defined in  [@sec:configuration-programming-languages] with the aims of providing composable, verifiable and validatable configuration files.
```

configuration (programming) languages are defined previous to this chapter.
What makes Nickel a programming language is discussed in the following sentences

> The language draws inspiration from existing projects such as Cue [@cue], Dhall [@Dhall] and most importantly Nix [@nix].
> Nickel implements a pure functional language with JSON-like data types and turing-complete lambda calculus.

Author: ```suggestion
Nickel implements a pure functional language with JSON-like data types and turing-complete lambda calculus.
The language draws inspiration from existing projects such as Cue [@cue], Dhall [@Dhall] and most importantly Nix [@nix].
```

Maybe switch these sentences around

However, Nickel sets itself apart from the existing projects by combining and advancing their strengths.
The language addresses concerns drawn from the experiences with Nix which employs a sophisticated modules system [@nixos-modules] to provide type-safe, composed (system) configuration files.
Nickel implements gradual type annotations, with runtime checked contracts to ensure even complex configurations remain correct.
Additionally, considering record merging on a language level facilitates modularization and composition of configurations.

> 

Author: ```suggestion

#### Nickel AST

Nickel's syntax tree is a single sum type, i.e., an enumeration of node types.
Each enumeration variant may refer to child nodes, representing a branch or hold terminal values in which case it is considered a leaf of the tree.
Additionally, tree nodes hold information about their position in the underlying code.

##### Basic Elements

The primitive values of the Nickel language are closely related to JSON.
On the leaf level, Nickel defines `Boolean`, `Number`, `String` and `Null`.
In addition to that the language implements native support for `Enum` values which are serialized as plain strings in less expressive formats such as JSON.
Each of these are terminal leafs in the syntax tree.

Completing JSON compatibility, `List` and `Record` constructs are present as well.
Records on a syntax level are Dictionaries, uniquely associating an identifier with a sub-node.

These data types constitute a static subset of Nickel which allows writing JSON compatible expressions as shown in [@lst:nickel-static].

```{.nickel #lst:nickel-static caption="Example of a static Nickel expression"}
{
  list = [ 1, "string", null],
  "some key" = "value"
} 
```

Beyond these basic elements, Nickel implements variables and functions as well as a special syntax for attaching metadata and recursive records.

##### Identifiers

The inclusion of variables to the language, implies an understanding of identifiers.
Such name bindings can be declared in multiple ways, e.g. `let` bindings, function arguments and records.
The usage of a name is always parsed as a single `Var` node wrapping the identifier.
Span information of identifiers is preserved by the parser and encoded in the `Ident` type. 

##### Variable Reference

```{.nickel #lst:nickel-let-binding caption="Let bindings and functions in nickel"}

// simple bindings
let name = <expr> in <expr>
let func = fun arg => <expr> in <expr>

// or with patterns
let name @ { field, with_default = 2 } = <expr> in <expr>
let func = fun arg @ { field, with_default = 2 } => 
  <expr> in 
  <expr>
```

Let bindings in their simplest form merely bind a name to a value expression and expose the name to the inner expression.
Hence, the `Let` node contains the binding and links to both implementation and scope subtrees.
The binding can be a simple name, a pattern or both by naming the pattern as shown in [@lst:nickel-let-binding].

```{.nickel #lst:nickel-args-function caption="Parsed representation of functions with multiple arguments"}
fun first second => first + second
// ...is parsed as
fun first =>
  fun second => first + second
```

Functions in Nickel are curried lambda expressions.
A function with multiple arguments gets broken down into nested single argument functions as seen in [@lst:nickel-args-function].
Function argument name binding therefore looks the same as in `let` bindings.

##### Meta Information

One key feature of Nickel is its gradual typing system [ref again?], which implies that values can be explicitly typed.
Complementing type information, it is possible to annotate values with contracts and additional metadata such as contracts, documentation, default values and merge priority using a special syntax as displayed in [@lst:nickel-meta].

```{.nickel #lst:nickel-meta caption="Example of a static Nickel expression"}
let Contract = { 
  foo | Num 
      | doc "I am foo",
  hello | Str
        | default = "world"
}
| doc "Just an example Contract"
in 
let value | #Contract = { foo = 9, }
in value == { foo = 9, hello = "world", } 

> true
```

Internally, the addition of annotations wraps the annotated term in a `MetaValue`, an additional tree node which describes its subtree. 
The expression shown in [@lst:nickel-meta-typed] translates to the AST in [@fig:nickel-meta-typed].

```{.nickel #lst:nickel-meta-typed caption="Example of a typed expression"}
let x: Num = 5 in x
```

```{.graphviz #fig:nickel-meta-typed caption="AST of typed expression" height=4.5cm}
strict digraph { 
  graph [fontname = "Fira Code"];
  node [fontname = "Fira Code"];
  edge [fontname = "Fira Code"];

  meta [label="MetaValue", color="green", shape="box"]
  let [label = "Let('x')"]
  num [label = "Num(5)"]
  var [label = "Var('x')"]

  meta -> let
  let -> num
  let -> var
}
```

##### Nested Record Access

Nickel supports both static and dynamic access to record fields.
If the field name is statically known, the access is said to be *static* accordingly.
Conversely, if the name requires evaluating a string from an expression the access is called *dynamic*.
An example is given in [@lst:nickel-static-dynamic]

```{.nickel #lst:nickel-static-dynamic caption="Examples for static and dynamic record access"}
let r = { foo = 1, "bar space" = 2} in
r.foo // static
r."bar space" // static
let field = "fo" ++ "o" in r."#{field}" // dynamic
```

The destruction of record fields is represented using a special set of AST nodes depending on whether the access is static or dynamic.
Static analysis does not evaluate dynamic fields and thus prevents the analysis of any deeper element starting with dynamic access.
Static access however can be used to resolve any intermediate reference.

Notably, Nickel represents static access chains in inverse order as unary operations which in turn puts the terminal `Var` node as a leaf in the tree.
[Figure @fig:nickel-static-access] shows the representation of the static access performed in [@lst:nickel-static-access] with the rest of the tree omitted.

```{.nickel #lst:nickel-static-access caption="Nickel static access"}
let x = {
  y = {
    z = 1,
  }
} in x.y.z
```

```{.graphviz #fig:nickel-static-access caption="AST of typed expression" height=6cm}
strict digraph { 
  graph [fontname = "Fira Code"];
  node [fontname = "Fira Code", margin=0.25];
  edge [fontname = "Fira Code"];

  rankdir="TD"

  let [label = "Let", color="grey"]
  rec [label = "omitted", color="grey", style="dashed", shape="box"]

  x [label = "Var('x')"]
  unop_x_y [label = ".y", shape = "triangle", margin=0.066]
  unop_y_z [label = ".z", shape = "triangle", margin=0.066]

  let -> rec
  let -> unop_y_z
  unop_y_z -> unop_x_y
  unop_x_y -> x
}
```

##### Record Shorthand

Nickel supports a shorthand syntax to efficiently define nested records similarly to how nested record fields are accessed.
As a comparison the example in [@lst:nickel-record-shorthand] uses the shorthand syntax which resolves to the semantically equivalent record defined in [@lst:nickel-record-no-shorthand]

```{.nickel #lst:nickel-record-shorthand caption="Nickel record defined using shorthand"}
{
  deeply.nested.record.field = true,
}
```

```{.nickel #lst:nickel-record-no-shorthand caption="Nickel record defined explicitly"}
{
  deeply = {
    nested = {
      record = { 
        field = true,
      }
    }
  }
}
```

Yet, on a syntax level Nickel generates a different representation.

```

#### Record Merging

> Nickel considers record merging as a fundamental operation that combines two records (i.e. JSON objects).

Reviewer: first define "record" in this context.

Author: ```suggestion
Nickel considers record merging as a fundamental operation that combines two records (i.e. JSON objects).
Records, as detailed in [@sec:basic-elements], are Nickel's native immutable dictionary structure.
```

  let [label = "Let('x')"]
  num [label = "Num(5)"]
  var [label = "Var('x')"]

  meta -> let
  let -> num
  let -> var
}
```

##### Nested Record Access

Nickel supports both static and dynamic access to record fields.
If the field name is statically known, the access is said to be *static* accordingly.
Conversely, if the name requires evaluating a string from an expression the access is called *dynamic*.
An example is given in [@lst:nickel-static-dynamic]

```{.nickel #lst:nickel-static-dynamic caption="Examples for static and dynamic record access"}
let r = { foo = 1, "bar space" = 2} in
r.foo // static
r."bar space" // static
let field = "fo" ++ "o" in r."#{field}" // dynamic
```

The destruction of record fields is represented using a special set of AST nodes depending on whether the access is static or dynamic.
Static analysis does not evaluate dynamic fields and thus prevents the analysis of any deeper element starting with dynamic access.
Static access however can be used to resolve any intermediate reference.

Notably, Nickel represents static access chains in inverse order as unary operations which in turn puts the terminal `Var` node as a leaf in the tree.

> [Figure @fig:nickel-static-access] shows the representation of the static access performed in [@lst:nickel-static-access] with the rest of the tree omitted.
> 
> ```{.nickel #lst:nickel-static-access caption="Nickel static access"}
> let x = {
>   y = {
>     z = 1,
>   }
> } in x.y.z
> ```
> 
> 
> ```{.graphviz #fig:nickel-static-access caption="AST of typed expression" height=6cm}
> strict digraph { 
>   graph [fontname = "Fira Code"];
>   node [fontname = "Fira Code", margin=0.25];
>   edge [fontname = "Fira Code"];
> 
>   rankdir="TD"
> 
>   let [label = "Let", color="grey"]
>   rec [label = "omitted", color="grey", style="dashed", shape="box"]
> 
>   x [label = "Var('x')"]
>   unop_x_y [label = ".y", shape = "triangle", margin=0.066]
>   unop_y_z [label = ".z", shape = "triangle", margin=0.066]
> 
> 
>   let -> rec
>   let -> unop_y_z
>   unop_y_z -> unop_x_y
>   unop_x_y -> x
> }
> ```
> 
> 
> 
> ##### Record Shorthand
> 
> Nickel supports a shorthand syntax to efficiently define nested records similarly to how nested record fields are accessed.
> As a comparison the example in [@lst:nickel-record-shorthand] uses the shorthand syntax which resolves to the semantically equivalent record defined in [@lst:nickel-record-no-shorthand]
> 
> ```{.nickel #lst:nickel-record-shorthand caption="Nickel record defined using shorthand"}
> {
>   deeply.nested.record.field = true,
> }
> ```
> 
> ```{.nickel #lst:nickel-record-no-shorthand caption="Nickel record defined explicitly"}
> {
>   deeply = {
>     nested = {
>       record = { 
>         field = true,
>       }
>     }
>   }
> }
> ```
> 
> Yet, on a syntax level Nickel generates a different representation.

Author: ```suggestion
```
